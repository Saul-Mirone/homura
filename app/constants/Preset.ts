export enum Preset {
  All = 'All Items',
  Archive = 'Archive',
  Starred = 'Starred',
  Unread = 'Unread',
}
