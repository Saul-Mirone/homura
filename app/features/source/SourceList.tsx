import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SideBar } from '../../components/SideBar';
import { Header } from '../../components/SideBar/Header';
import { SideBarItem } from '../../components/SideBar/SideBarItem';
import { AppDispatch } from '../../store';
import {
  fetchSources,
  selectSourceList,
  setCurrentSource,
  unsubscribeById,
  updateSourceById,
} from './sourceSlice';

export const SourceList: React.FC<{ bottom: JSX.Element }> = ({ bottom }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { list, activeId, mode, totalCount } = useSelector(selectSourceList);

  React.useEffect(() => {
    dispatch(fetchSources(mode));
  }, [dispatch, mode]);

  const overview = (
    <Header
      mode={mode}
      active={activeId}
      count={totalCount}
      onClick={(id) => dispatch(setCurrentSource(id))}
    />
  );

  return (
    <SideBar
      onClick={() => dispatch(setCurrentSource(null))}
      overview={overview}
      bottom={bottom}
    >
      {list.map(({ id, name, count, icon }) => (
        <SideBarItem
          key={id.toString()}
          id={id}
          url={icon}
          name={name}
          count={count}
          active={id === activeId}
          onConfirmModify={(nextName) =>
            dispatch(updateSourceById({ id, name: nextName }))
          }
          onUnsubscribe={() => dispatch(unsubscribeById(id))}
          onClick={() => dispatch(setCurrentSource(id))}
        />
      ))}
    </SideBar>
  );
};
