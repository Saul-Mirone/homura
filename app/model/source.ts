import {
  Association,
  DataTypes,
  HasManyAddAssociationMixin,
  HasManyCountAssociationsMixin,
  HasManyCreateAssociationMixin,
  HasManyGetAssociationsMixin,
  HasManyHasAssociationMixin,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import { Post } from './post';

export interface SourceAttributes {
  id: number;
  name: string;
  link: string;
  sourceUrl: string;
  icon: string | null;
}

export type CreateSourceAttributes = Optional<SourceAttributes, 'id'>;

export class Source
  extends Model<SourceAttributes, CreateSourceAttributes>
  implements SourceAttributes {
  public id!: number;

  public name!: string;

  public link!: string;

  public sourceUrl!: string;

  public icon!: string | null;

  public getPosts!: HasManyGetAssociationsMixin<Post>;

  public addPost!: HasManyAddAssociationMixin<Post, number>;

  public hasPost!: HasManyHasAssociationMixin<Post, number>;

  public countPosts!: HasManyCountAssociationsMixin;

  public createPost!: HasManyCreateAssociationMixin<Post>;

  public posts?: Post[];

  public toJSON!: () => SourceAttributes & {
    posts: Array<ReturnType<Post['toJSON']>>;
  };

  public static associations: {
    posts: Association<Source, Post>;
  };
}

export const initSource = (sequelize: Sequelize): void => {
  Source.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceUrl: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      icon: {
        type: DataTypes.STRING,
        allowNull: true,
      },
      link: {
        type: DataTypes.STRING,
        allowNull: false,
      },
    },
    { sequelize, tableName: 'sources' }
  );

  Source.hasMany(Post, {
    sourceKey: 'id',
    foreignKey: 'sourceId',
    as: 'posts',
  });
  Post.belongsTo(Source, {
    targetKey: 'id',
    foreignKey: 'sourceId',
    as: 'source',
  });
};
