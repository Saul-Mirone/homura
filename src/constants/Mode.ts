export enum Mode {
    All = 'All',
    Unread = 'Unread',
    Starred = 'Starred',
}
