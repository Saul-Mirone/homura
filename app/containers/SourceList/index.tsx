import React from 'react';
import { Creator } from '../../features/creator/Creator';
import { Source } from '../../features/source/Source';

export const SourceList: React.FC = () => {
  return <Source bottom={<Creator />} />;
};
