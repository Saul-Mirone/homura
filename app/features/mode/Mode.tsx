import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Toolkit } from '../../components/SubSideBar/ToolKit';
import { AppDispatch } from '../../store';
import { selectMode, switchMode } from './modeSlice';

export const Mode: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const mode = useSelector(selectMode);

  return <Toolkit mode={mode} onSwitchMode={(m) => dispatch(switchMode(m))} />;
};
