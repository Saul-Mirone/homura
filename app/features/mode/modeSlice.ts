import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Mode } from '../../constants/Mode';
import type { RootState } from '../../store';

const modeSlice = createSlice({
  name: 'mode',
  initialState: Mode.All,
  reducers: {
    switchMode: (_, action: PayloadAction<Mode>) => action.payload,
  },
});

export const { switchMode } = modeSlice.actions;

export const modeReducer = modeSlice.reducer;

export const selectMode = (state: RootState) => state.mode;
