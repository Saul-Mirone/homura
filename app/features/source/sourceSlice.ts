import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import { Step } from '../../components/SideBar/BottomBar';
import { Mode } from '../../constants/Mode';
import { Preset } from '../../constants/Preset';
import { Status } from '../../constants/Status';
import type { RootState } from '../../store';

type SourceItem = {
  id: number;
  name: string;
  link: string;
  count: number;
  icon: string | null;
};

type SourceListState = {
  activeId: number | Preset | null;
  list: SourceItem[];
  fetchListStatus: Status;
};

type SourceSubscribeState = {
  subscribeLink: string;
  subscribeName: string;
  subscribeStep: Step | null;
  subscribeStatus: Status;
  subscribeError: null | string;
};

export type State = SourceListState & SourceSubscribeState;

const initialSourceList: SourceListState = {
  list: [],
  activeId: null,
  fetchListStatus: Status.Idle,
};

const initialSourceSubscribeState: SourceSubscribeState = {
  subscribeLink: '',
  subscribeName: '',
  subscribeStep: null,
  subscribeStatus: Status.Idle,
  subscribeError: null,
};

const initialState: State = {
  ...initialSourceList,
  ...initialSourceSubscribeState,
};

const fetchSourceListAPI = async (mode: Mode) => {
  const countType = mode === Mode.Starred ? 'starred' : 'unread';
  const data = await channel.getSourceList(countType);
  return data.map(({ id, name, icon, link, count }) => ({
    id,
    name,
    icon,
    link,
    count,
  }));
};

export const fetchSources = createAsyncThunk(
  'source/fetchSources',
  (mode: Mode) => {
    return fetchSourceListAPI(mode);
  }
);

export const syncSources = createAsyncThunk(
  'source/syncSources',
  async (mode: Mode) => {
    await channel.sync();
    return fetchSourceListAPI(mode);
  }
);

export const unsubscribeById = createAsyncThunk(
  'source/unsubscribeById',
  async (id: number) => {
    await channel.removeSourceById(id);
    return id;
  }
);

export const updateSourceById = createAsyncThunk(
  'source/updateSourceById',
  async ({ id, name }: { id: number; name: string }) => {
    await channel.updateSourceNameById(id, name);

    return { id, name };
  }
);

export const searchUrlForSource = createAsyncThunk(
  'source/searchUrlForSource',
  async (link: string) => {
    const result = await channel.checkUrl(link);
    if (!result) {
      throw new Error('ParseRSSFailed');
    }
    return result;
  }
);

export const subscribeToSource = createAsyncThunk(
  'source/subscribeToSource',
  async ({ name, mode }: { name: string; mode: Mode }) => {
    const { posts, icon, link, id } = await channel.confirm(name);

    return {
      id,
      name,
      link,
      icon,
      count: mode === Mode.Starred ? 0 : posts.length,
    };
  }
);

const sourceSlice = createSlice({
  name: 'source',
  initialState,
  reducers: {
    setCurrentSource: (
      state,
      action: PayloadAction<number | Preset | null>
    ) => {
      state.activeId = action.payload;
    },
    showSubscribeBar: (state) => {
      state.subscribeStep = Step.EnterUrl;
      state.subscribeStatus = Status.Idle;
    },
    resetSubscribeError: (state) => {
      state.subscribeError = null;
    },
    resetSubscribeState: (state) => {
      state.subscribeStep = null;
      state.subscribeError = null;
      state.subscribeLink = '';
      state.subscribeName = '';
      state.subscribeStatus = Status.Idle;
    },

    // TODO: need to be reviewed
    countDownOne: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count -= 1;
    },

    // TODO: need to be reviewed
    countUpOne: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count += 1;
    },

    // TODO: need to be reviewed
    countToZero: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count = 0;
    },
  },
  extraReducers: (builder) =>
    builder
      .addCase(fetchSources.pending, (state) => {
        state.fetchListStatus = Status.Pending;
      })
      .addCase(fetchSources.fulfilled, (state, action) => {
        state.fetchListStatus = Status.Succeeded;
        state.list = action.payload;
      })
      .addCase(fetchSources.rejected, (state) => {
        state.fetchListStatus = Status.Failed;
      })
      .addCase(syncSources.pending, (state) => {
        state.fetchListStatus = Status.Pending;
      })
      .addCase(syncSources.fulfilled, (state, action) => {
        state.fetchListStatus = Status.Succeeded;
        state.list = action.payload;
      })
      .addCase(syncSources.rejected, (state) => {
        state.fetchListStatus = Status.Failed;
      })
      .addCase(unsubscribeById.fulfilled, (state, action) => {
        state.activeId = null;
        state.list = state.list.filter((x) => x.id !== action.payload);
      })
      .addCase(updateSourceById.fulfilled, (state, action) => {
        state.activeId = null;
        const target = state.list.find((x) => x.id === action.payload.id);
        if (!target) return;
        target.name = action.payload.name;
      })
      .addCase(searchUrlForSource.pending, (state, action) => {
        state.subscribeLink = action.meta.arg;
        state.subscribeStatus = Status.Pending;
      })
      .addCase(searchUrlForSource.fulfilled, (state, action) => {
        state.subscribeName = action.payload;
        state.subscribeStep = Step.EnterName;
      })
      .addCase(searchUrlForSource.rejected, (state, action) => {
        state.subscribeStatus = Status.Failed;
        state.subscribeError = action.error.message ?? 'Unknown Error';
      })
      .addCase(subscribeToSource.fulfilled, (state, action) => {
        const { id } = action.payload;
        state.subscribeStatus = Status.Succeeded;
        state.list.push(action.payload);
        state.activeId = id;
        state.subscribeStep = null;
        state.subscribeError = null;
        state.subscribeLink = '';
        state.subscribeName = '';
      }),
});

export const {
  setCurrentSource,
  resetSubscribeError,
  resetSubscribeState,
  showSubscribeBar,

  countDownOne,
  countUpOne,
  countToZero,
} = sourceSlice.actions;

export const sourceReducer = sourceSlice.reducer;

export const selectSourceList = (state: RootState) => {
  const { source, mode } = state;
  const totalCount = source.list.reduce((acc, cur) => acc + cur.count, 0);
  const { list, activeId } = source;

  return {
    activeId,
    totalCount,
    mode,
    list: list.filter((x) => {
      if (mode === Mode.All || activeId === x.id) return true;
      return x.count > 0;
    }),
  };
};

export const selectSourceOperation = (state: RootState) => {
  const { source, mode } = state;
  const {
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeStatus,
    subscribeError,
    fetchListStatus,
  } = source;
  const loading =
    subscribeStatus === Status.Pending || fetchListStatus === Status.Pending;
  return {
    mode,
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeError,
    loading,
  };
};
