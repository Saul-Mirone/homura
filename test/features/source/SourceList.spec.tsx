/* eslint-disable @typescript-eslint/no-non-null-assertion */

// organize-imports-ignore
// make sure mock file on the top of imports
import { mockChannel } from '../../test-tools/mockChannel';

import { configureStore } from '@reduxjs/toolkit';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { Mode } from '../../../src/constants/Mode';
import { Status } from '../../../src/constants/Status';
import * as modeSlice from '../../../src/features/mode/modeSlice';
import { SourceList } from '../../../src/features/source/SourceList';
import * as sourceSlice from '../../../src/features/source/sourceSlice';
import { Preset } from '../../../src/constants/Preset';
import userEvent from '@testing-library/user-event';

const mockAppend = jest.fn();
const mockPopup = jest.fn();

jest.mock('electron', () => ({
    remote: {
        Menu: jest.fn(() => ({
            append: mockAppend,
            popup: mockPopup,
        })),
        MenuItem: jest.fn(({ label, click }) => ({
            label,
            click,
        })),
    },
}));

function setup(
    preloadedState: {
        source: Partial<sourceSlice.State>;
        mode?: Mode;
    } = {
        source: {},
        mode: Mode.All,
    },
) {
    const initialSourceState = {
        list: [],
        activeId: undefined,
        fetchListStatus: Status.Idle,
        syncListStatus: Status.Idle,
        subscribeLink: '',
        subscribeName: '',
        subscribeStep: undefined,
        subscribeStatus: Status.Idle,
        subscribeError: undefined,
    };
    const state: {
        source: sourceSlice.State;
        mode: Mode;
    } = {
        source: { ...initialSourceState, ...preloadedState.source },
        mode: preloadedState.mode ?? Mode.All,
    };
    const store = configureStore({
        reducer: { source: sourceSlice.sourceReducer, mode: modeSlice.modeReducer },
        preloadedState: state,
    });

    const utils = render(
        <Provider store={store}>
            <SourceList bottom={<div />} />
        </Provider>,
    );

    return {
        ...utils,
        store,
        el: {
            list: utils.getByRole('menu'),
            get items() {
                return utils.getAllByRole('menuitem');
            },
            get totalCount() {
                return Number(utils.container.querySelector('.sidebar-overview-item__count')!.innerHTML);
            },
        },
    };
}

type Menu = {
    label: string;
    click: () => void;
};

function prepareChannel() {
    mockChannel.getSourceList.mockResolvedValue([
        {
            id: 1,
            name: 'data-1',
            count: 24,
            link: 'link-1',
        },
        {
            id: 2,
            name: 'data-2',
            count: 0,
            icon: 'fake-icon-url',
            link: 'link-2',
        },
        {
            id: 3,
            name: 'data-3',
            count: 2,
            link: 'link-3',
        },
    ]);
    mockChannel.sync.mockResolvedValue();
    mockChannel.removeSourceById.mockResolvedValue();
    mockChannel.updateSourceNameById.mockResolvedValue();
}

function prepareMenu() {
    let menu: Array<Menu> = [];
    let menuBuffer: Array<Menu> = [];
    mockAppend.mockImplementation((x) => menuBuffer.push(x));
    mockPopup.mockImplementation(() => {
        menu = [...menuBuffer];
        menuBuffer = [];
    });

    return () => menu;
}

test('SourceList', async () => {
    prepareChannel();

    const getMenu = prepareMenu();

    const syncSourcesSpy = jest.spyOn(sourceSlice, 'syncSources');
    const unsubscribeSpy = jest.spyOn(sourceSlice, 'unsubscribeById');
    const fetchSourcesSpy = jest.spyOn(sourceSlice, 'fetchSources');
    const setCurrentSourceSpy = jest.spyOn(sourceSlice, 'setCurrentSource');
    const updateSourceByIdSpy = jest.spyOn(sourceSlice, 'updateSourceById');

    // setup
    const { el, container } = setup();

    // first init
    expect(el.list).toMatchSnapshot();
    expect(el.items).toHaveLength(2);
    expect(syncSourcesSpy).toBeCalledTimes(1);

    await waitFor(() => expect(fetchSourcesSpy).toHaveBeenCalledTimes(1));

    expect(fetchSourcesSpy).toBeCalledWith(Mode.All);
    expect(el.items).toHaveLength(5);
    expect(el.list).toMatchSnapshot();

    expect(el.totalCount).toBe(26);

    // click on item
    const data1Item = el.items[2]!;

    fireEvent.click(data1Item);
    expect(setCurrentSourceSpy).toBeCalledWith(1);
    expect(data1Item).toHaveClass('active');

    expect(container.querySelectorAll('.sidebar-item__count')).toHaveLength(2);
    expect(container.querySelector('.sidebar-item__count')!.innerHTML).toBe('24');

    // click on preset
    const presetAll = el.items[0]!;

    fireEvent.click(presetAll);
    expect(setCurrentSourceSpy).toBeCalledWith(Preset.All);
    expect(data1Item).not.toHaveClass('active');
    expect(presetAll).toHaveClass('active');

    // click on empty area
    fireEvent.click(el.list);
    expect(setCurrentSourceSpy).toBeCalledWith();
    expect(presetAll).not.toHaveClass('active');

    const count = setCurrentSourceSpy.mock.calls.length;
    fireEvent.click(el.list);
    expect(setCurrentSourceSpy).toBeCalledTimes(count);

    // context menu > unsubscribe
    fireEvent.contextMenu(data1Item);
    expect(mockPopup).toBeCalledTimes(1);
    expect(getMenu().map((x) => x.label)).toEqual(['unsubscribe', 'edit']);

    await act(() => getMenu()[0]!.click());

    expect(unsubscribeSpy).toBeCalledWith(1);
    expect(el.list).toMatchSnapshot();
    expect(el.totalCount).toBe(2);
    expect(el.items).toHaveLength(4);

    // context menu > edit
    const data2Item = el.items[2]!;
    fireEvent.contextMenu(data2Item);
    act(() => getMenu()[1]!.click());

    const input = container.querySelector('input')!;
    const button = container.querySelector('.sidebar-item__confirm-icon')!;

    expect(input).toHaveValue('data-2');

    userEvent.type(input, '-new-value');
    expect(input).toHaveValue('data-2-new-value');

    fireEvent.click(button);

    await waitFor(() => expect(updateSourceByIdSpy).toHaveBeenCalledTimes(1));
    expect(updateSourceByIdSpy).toBeCalledWith({
        id: 2,
        name: 'data-2-new-value',
    });
    expect(data2Item).toContainHTML('data-2-new-value');
    expect(el.list).toMatchSnapshot();
});
