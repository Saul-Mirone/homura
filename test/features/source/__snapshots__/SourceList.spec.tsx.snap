// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceList 1`] = `
<div
  aria-label="Sources"
  class="sidebar-container"
  role="menu"
  tabindex="0"
>
  <div
    class="relative flex-1 thin-scroll"
  >
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 10h16M4 14h16M4 18h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          All Items
        </div>
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 8h14M5 8a2 2 0 110-4h14a2 2 0 110 4M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8m-9 4h4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          Archive
        </div>
      </div>
    </div>
    <div
      class="pt-10"
    />
  </div>
  <div />
</div>
`;

exports[`SourceList 2`] = `
<div
  aria-label="Sources"
  class="sidebar-container"
  role="menu"
  tabindex="0"
>
  <div
    class="relative flex-1 thin-scroll"
  >
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 10h16M4 14h16M4 18h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          All Items
        </div>
      </div>
      <div
        class="sidebar-overview-item__count"
      >
        26
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 8h14M5 8a2 2 0 110-4h14a2 2 0 110 4M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8m-9 4h4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          Archive
        </div>
      </div>
    </div>
    <div
      class="pt-10"
    />
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-1
        </div>
      </div>
      <div
        class="sidebar-item__count"
      >
        24
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="image-icon"
        >
          <div>
            <img
              alt="icon"
              src="fake-icon-url"
            />
          </div>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-2
        </div>
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-3
        </div>
      </div>
      <div
        class="sidebar-item__count"
      >
        2
      </div>
    </div>
  </div>
  <div />
</div>
`;

exports[`SourceList 3`] = `
<div
  aria-label="Sources"
  class="sidebar-container"
  role="menu"
  tabindex="0"
>
  <div
    class="relative flex-1 thin-scroll"
  >
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 10h16M4 14h16M4 18h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          All Items
        </div>
      </div>
      <div
        class="sidebar-overview-item__count"
      >
        2
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 8h14M5 8a2 2 0 110-4h14a2 2 0 110 4M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8m-9 4h4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          Archive
        </div>
      </div>
    </div>
    <div
      class="pt-10"
    />
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="image-icon"
        >
          <div>
            <img
              alt="icon"
              src="fake-icon-url"
            />
          </div>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-2
        </div>
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-3
        </div>
      </div>
      <div
        class="sidebar-item__count"
      >
        2
      </div>
    </div>
  </div>
  <div />
</div>
`;

exports[`SourceList 4`] = `
<div
  aria-label="Sources"
  class="sidebar-container"
  role="menu"
  tabindex="0"
>
  <div
    class="relative flex-1 thin-scroll"
  >
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 10h16M4 14h16M4 18h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          All Items
        </div>
      </div>
      <div
        class="sidebar-overview-item__count"
      >
        2
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-overview-item"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 8h14M5 8a2 2 0 110-4h14a2 2 0 110 4M5 8v10a2 2 0 002 2h10a2 2 0 002-2V8m-9 4h4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="sidebar-overview-item__text"
        >
          Archive
        </div>
      </div>
    </div>
    <div
      class="pt-10"
    />
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="image-icon"
        >
          <div>
            <img
              alt="icon"
              src="fake-icon-url"
            />
          </div>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-2-new-value
        </div>
      </div>
    </div>
    <div
      class=" rss-item"
      role="menuitem"
      tabindex="0"
    >
      <div
        class="sidebar-item__container"
      >
        <div
          class="logo-icon"
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          class="ml-2 text-xs truncate"
        >
          data-3
        </div>
      </div>
      <div
        class="sidebar-item__count"
      >
        2
      </div>
    </div>
  </div>
  <div />
</div>
`;
